File: settlement/room/spirit/shrine/ROOM_SHRINE.java
package settlement.room.spirit.shrine;

import java.io.IOException;

import init.religion.RELIGIONS;
import init.religion.Religion;
import init.type.NEEDS;
import settlement.misc.util.FSERVICE;
import settlement.path.finders.SFinderRoomService;
import settlement.room.main.RoomBlueprintIns;
import settlement.room.main.category.RoomCategorySub;
import settlement.room.main.furnisher.Furnisher;
import settlement.room.main.util.RoomInitData;
import settlement.room.service.module.RoomService;
import settlement.room.service.module.RoomService.ROOM_SERVICE_HASER;
import snake2d.util.file.FileGetter;
import snake2d.util.file.FilePutter;

public final class ROOM_SHRINE extends RoomBlueprintIns<ShrineInstance> implements ROOM_SERVICE_HASER {

	public final Religion religion;
	final RoomService data; 
	
	final Constructor constructor;
	final Service bed;
	
	public ROOM_SHRINE(String key, int index, RoomInitData init, RoomCategorySub block) throws IOException {
		super(index, init, key, block);
		
		religion = RELIGIONS.MAP().read(init.data());
		bed = new Service(this);
		data = new RoomService(this, init, NEEDS.TYPES().SHRINE) {
			
			@Override
			public FSERVICE service(int tx, int ty) {
				return bed.get(tx, ty);
			}
		};
		constructor = new Constructor(this, init);
	}
	
	@Override
	protected void update(double ds) {
		// TODO Auto-generated method stub
		
	}
	
	public Service bed(int tx, int ty) {
		return bed.get(tx, ty);
	}
	
	@Override
	public Furnisher constructor() {
		return constructor;
	}
	
	@Override
	public SFinderRoomService service(int tx, int ty) {
		return data.finder;
	}

	@Override
	protected void saveP(FilePutter saveFile){
		data.saver.save(saveFile);
	}
	
	@Override
	protected void loadP(FileGetter saveFile) throws IOException{
		data.saver.load(saveFile);
	}
	
	@Override
	protected void clearP() {
		data.saver.clear();
	}
	
	@Override
	public RoomService service() {
		return data;
	}

}

File: LegacySource/settlement/room/spirit/shrine/ROOM_SHRINE.cs
using System;
using System.IO;

using init.religion;
using init.type;
using settlement.misc.util;
using settlement.path.finders;
using settlement.room.main;
using settlement.room.main.category;
using settlement.room.main.furnisher;
using settlement.room.main.util;
using settlement.room.service.module;
using snake2d.util.file;

public sealed class ROOM_SHRINE : RoomBlueprintIns<ShrineInstance>, ROOM_SERVICE_HASER
{
    public readonly Religion religion;
    private readonly RoomService data;

    private readonly Constructor constructor;
    private readonly Service bed;

    public ROOM_SHRINE(string key, int index, RoomInitData init, RoomCategorySub block) : base(index, init, key, block)
    {
        religion = RELIGIONS.MAP().Read(init.data());
        bed = new Service(this);
        data = new RoomService(this, init, NEEDS.TYPES().SHRINE)
        {
            public FSERVICE Service(int tx, int ty)
            {
                return bed.Get(tx, ty);
            }
        };
        constructor = new Constructor(this, init);
    }

    protected override void Update(double ds)
    {
        // TODO Auto-generated method stub
    }

    public Service Bed(int tx, int ty)
    {
        return bed.Get(tx, ty);
    }

    public Furnisher Constructor()
    {
        return constructor;
    }

    public SFinderRoomService Service(int tx, int ty)
    {
        return data.Finder;
    }

    protected override void SaveP(FilePutter saveFile)
    {
        data.Saver.Save(saveFile);
    }

    protected override void LoadP(FileGetter saveFile)
    {
        data.Saver.Load(saveFile);
    }

    protected override void ClearP()
    {
        data.Saver.Clear();
    }

    public RoomService Service()
    {
        return data;
    }
}